File: julia/src/benchmarks/BrainfuckArray.jl
mutable struct BrainfuckArray <: AbstractBenchmark
    program_text::String
    warmup_text::String
    result::UInt32

    function BrainfuckArray()
        program_text = Helper.config_s("BrainfuckArray", "program")
        warmup_text = Helper.config_s("BrainfuckArray", "warmup_program")
        new(program_text, warmup_text, UInt32(0))
    end
end

name(b::BrainfuckArray)::String = "BrainfuckArray"

mutable struct Tape
    data::Vector{UInt8}
    pos::Int

    function Tape()

        data = Vector{UInt8}(undef, 30000)
        fill!(data, 0x00)

        new(data, 1)
    end
end

function get(tape::Tape)::UInt8
    return tape.data[tape.pos]
end

function inc!(tape::Tape)
    tape.data[tape.pos] += 0x01
end

function dec!(tape::Tape)
    tape.data[tape.pos] -= 0x01
end

function advance!(tape::Tape)
    tape.pos += 1
    if tape.pos > length(tape.data)
        push!(tape.data, 0x00)
    end
end

function devance!(tape::Tape)
    if tape.pos > 1
        tape.pos -= 1
    end
end

struct Program
    commands::Vector{UInt8}
    jumps::Vector{Int}

    function Program(text::String)

        len = length(text)
        temp = Vector{UInt8}(undef, len)
        count = 0

        for c in text
            if c in "[]<>+-,."
                count += 1
                temp[count] = UInt8(c)
            end
        end

        commands = Vector{UInt8}(undef, count)
        copyto!(commands, 1, temp, 1, count)

        jumps = zeros(Int, count)
        stack = Vector{Int}(undef, count)
        sp = 0

        for i = 1:count
            cmd = commands[i]
            if cmd == UInt8('[')
                sp += 1
                stack[sp] = i
            elseif cmd == UInt8(']')
                if sp > 0
                    start_idx = stack[sp]
                    sp -= 1
                    jumps[start_idx] = i
                    jumps[i] = start_idx
                end
            end
        end

        new(commands, jumps)
    end
end

function run(prog::Program)::Int64
    result = Int64(0)
    tape = Tape()
    pc = 1
    cmds = prog.commands
    jmps = prog.jumps

    while pc <= length(cmds)
        cmd = cmds[pc]

        if cmd == UInt8('+')
            inc!(tape)
        elseif cmd == UInt8('-')
            dec!(tape)
        elseif cmd == UInt8('>')
            advance!(tape)
        elseif cmd == UInt8('<')
            devance!(tape)
        elseif cmd == UInt8('[')
            if get(tape) == 0x00
                pc = jmps[pc]
                continue
            end
        elseif cmd == UInt8(']')
            if get(tape) != 0x00
                pc = jmps[pc]
                continue
            end
        elseif cmd == UInt8('.')
            result = (result << 2) + Int64(get(tape))
        end

        pc += 1
    end

    return result
end

function _run(text::String)::Int64
    return run(Program(text))
end

function warmup(b::BrainfuckArray)
    warmup_iters = warmup_iterations(b)
    for _ = 1:warmup_iters
        _run(b.warmup_text)
    end
end

function run(b::BrainfuckArray, iteration_id::Int64)
    run_result = _run(b.program_text)
    b.result += Helper.to_u32(run_result)
end

checksum(b::BrainfuckArray)::UInt32 = b.result

File: nim/src/benchmarks/brainfuckarray.nim
import std/[strutils, sequtils]
import ../benchmark
import ../helper

type
  Tape = object
    tape: seq[uint8]
    pos: int

  Program = object
    commands: seq[uint8]
    jumps: seq[int]

proc newTape(): Tape =
  result = Tape(tape: newSeq[uint8](30000), pos: 0)

proc get(self: Tape): uint8 = self.tape[self.pos]
proc inc(self: var Tape) = self.tape[self.pos].inc
proc dec(self: var Tape) = self.tape[self.pos].dec

proc advance(self: var Tape) =
  self.pos.inc
  if self.pos >= self.tape.len:
    self.tape.add(0)

proc devance(self: var Tape) =
  if self.pos > 0:
    self.pos.dec

proc newProgram(text: string): Program =
  var commands: seq[uint8] = @[]
  for c in text:
    if c in "[]<>+-.,":
      commands.add(c.uint8)

  var jumps = newSeq[int](commands.len)
  var stack: seq[int] = @[]

  for i, cmd in commands:
    if cmd == '['.uint8:
      stack.add(i)
    elif cmd == ']'.uint8 and stack.len > 0:
      let start = stack.pop()
      jumps[start] = i
      jumps[i] = start

  Program(commands: commands, jumps: jumps)

proc run(self: Program): int64 =
  var tape = newTape()
  var pc = 0
  result = 0

  while pc < self.commands.len:
    let cmd = self.commands[pc]
    if cmd == '+'.uint8:
      tape.inc()
    elif cmd == '-'.uint8:
      tape.dec()
    elif cmd == '>'.uint8:
      tape.advance()
    elif cmd == '<'.uint8:
      tape.devance()
    elif cmd == '['.uint8:
      if tape.get == 0:
        pc = self.jumps[pc]
        continue
    elif cmd == ']'.uint8:
      if tape.get != 0:
        pc = self.jumps[pc]
        continue
    elif cmd == '.'.uint8:
      result = (result shl 2) + tape.get.int64
    pc.inc

type
  BrainfuckArray* = ref object of Benchmark
    programText: string
    warmupText: string
    resultVal: uint32

proc newBrainfuckArray(): Benchmark =
  BrainfuckArray()

method name(self: BrainfuckArray): string = "BrainfuckArray"

method prepare(self: BrainfuckArray) =
  self.programText = config_s(self.name, "program")
  self.warmupText = config_s(self.name, "warmup_program")
  self.resultVal = 0

proc runProgram(text: string): int64 =
  let program = newProgram(text)
  program.run()

method warmup(self: BrainfuckArray) =
  let prepare_iters = self.warmup_iterations
  for _ in 0..<prepare_iters:
    discard runProgram(self.warmupText)

method run(self: BrainfuckArray, iteration_id: int) =
  let run_result = runProgram(self.programText)
  self.resultVal += run_result.uint32

method checksum(self: BrainfuckArray): uint32 =
  self.resultVal

registerBenchmark("BrainfuckArray", newBrainfuckArray)
